File: src/lib/gmp_lib/src/GMP/GMPo_Update.cpp
#include <gmp_lib/GMP/GMPo_Update.h>

// N-DoF GMP Update class
//

#include <gmp_lib/GMP/GMPo.h>
#include <gmp_lib/GMP/GMP_Update.h>
#include <gmp_lib/math/quaternions.h>

namespace as64_
{

namespace gmp_
{

  GMPo_Update::GMPo_Update(GMPo *gmp): GMP_Update(gmp)
  {
    this->gmp_o = gmp;
  }

  void GMPo_Update::updateQuat(double x, const arma::vec &Q, double r_n)
  {
    arma::vec y = GMPo::quat2q(Q, this->gmp_o->getQ0());
    updatePos(x, y, r_n);
  }

  void GMPo_Update::updateRotVel(double x, double x_dot, const arma::vec &rotVel, const arma::vec &Q, double r_n)
  {
    arma::vec Q1 = GMPo::getQ1(Q, this->gmp_o->getQ0());
    arma::vec y_dot = gmp_::rotVel_to_qLogDot(rotVel, Q1);
    updateVel(x, x_dot, y_dot, r_n);
  }

  void GMPo_Update::updateRotAccel(double x, double x_dot, double x_ddot, const arma::vec &rotAccel, const arma::vec &rotVel, const arma::vec &Q, double r_n)
  {
    arma::vec Q1 = GMPo::getQ1(Q, this->gmp_o->getQ0());
    arma::vec y_ddot = gmp_::rotAccel_to_qLogDDot(rotAccel, rotVel, Q1);
    updateAccel(x, x_dot, x_ddot, y_ddot, r_n);
  }

} // namespace gmp_

} // namespace as64_

File: src/matlab/lib/gmp_lib/GMPo/GMPo_Update.m
%% N-DoF GMP Update class
%

classdef GMPo_Update < GMP_Update

    methods (Access = public)
        
        %% GMP constructor.
        %  @param[in] gmp: n_DoF dmp.
        function this = GMPo_Update(gmp)
                
            if ( ~ strcmpi(class(gmp),'GMPo') )
                error(['[GMPo_Update::GMPo_Update]: argument of type ''GMPo'' was expected, but a type ''' class(gmp) ''' was provided instead']);
            end
                
            this@GMP_Update(gmp);

        end

        % function initSigmaw(this)
        
        % function initSigmaWfromMsr(this, x_data)
        
        % function enableSigmawUpdate(this, flag)
        
        % function setSigmaW(this, Sw)
        
        % function Sw = getSigmaW(this)
        
        % function setMsrNoiseVar(this, rv)
        
        
        %% ==================================================
        %% ===============   Online update  =================
        %% ==================================================
        
        % function updatePos(this, x, y, r_n)
        
        % function updateVel(this, x, x_dot, y_dot, r_n)
        
        % function updateAccel(this, x, x_dot, x_ddot, y_ddot, r_n)
        
        function updateQuat(this, x, Q, r_n)
            
            if (nargin < 4), r_n=this.rv; end
            
            y = GMPo.quat2q(Q, this.gmp.getQ0());
            this.updatePos(x, y, r_n);
            
        end
        
        function updateRotVel(this, x, x_dot, rotVel, Q, r_n)
            
            if (nargin < 6), r_n=this.rv; end
            
            Q1 = GMPo.getQ1(Q, this.gmp.getQ0());
            y_dot = gmp_.rotVel_to_qLogDot(rotVel, Q1);
            this.updateVel(x, x_dot, y_dot, r_n);
            
        end
        
        function updateRotAccel(this, x, x_dot, x_ddot, rotAccel, rotVel, Q, r_n)
            
            if (nargin < 8), r_n=this.rv; end
            
            Q1 = GMPo.getQ1(Q, this.gmp.getQ0());
            y_ddot = gmp_.rotAccel_to_qLogDDot(rotAccel, rotVel, Q1);
            this.updateAccel(x,x_dot,x_ddot, y_ddot, r_n);
          
        end

        % function updateWeights(this, s, Z, type, r_n)
        
        % function plotWeightsCovariance(this, ax)
    end
    
end
